pkg/registry/node/etcd/etcd.go

```diff
@@ -124,30 +125,41 @@ func (r *REST) ResourceLocation(ctx api.Context, id string) (*url.URL, http.Roun
 
 var _ = client.ConnectionInfoGetter(&REST{})
 
-func (r *REST) getKubeletPort(ctx api.Context, nodeName string) (int, error) {
-	// We probably shouldn't care about context when looking for Node object.
-	obj, err := r.Get(ctx, nodeName)
+func (r *REST) getKubeletHostPortPair(nodeName string) (string, int, error) {
+	obj, err := r.Get(api.NewDefaultContext(), nodeName)
 	if err != nil {
-		return 0, err
+		return "", 0, err
 	}
 	node, ok := obj.(*api.Node)
 	if !ok {
-		return 0, fmt.Errorf("Unexpected object type: %#v", node)
+		return "", 0, fmt.Errorf("Unexpected object type: %#v", node)
 	}
-	return int(node.Status.DaemonEndpoints.KubeletEndpoint.Port), nil
+	port := int(node.Status.DaemonEndpoints.KubeletEndpoint.Port)
+	return getNodeAddress(node), port, nil
 }
 
-func (c *REST) GetConnectionInfo(ctx api.Context, nodeName string) (string, uint, http.RoundTripper, error) {
-	scheme, port, transport, err := c.connection.GetConnectionInfo(ctx, nodeName)
-	if err != nil {
-		return "", 0, nil, err
+func getNodeAddress(node *api.Node) string {
+	nodeAddresses := node.Status.Addresses
+	for _, address := range nodeAddresses {
+		if address.Type == api.NodeHostName {
+			return address.Address
+		}
 	}
-	daemonPort, err := c.getKubeletPort(ctx, nodeName)
+	nodeName := node.Name
+	glog.Warningf("Failed to retrieve Hostname for Node %s, falling back to NodeName", nodeName)
+	return nodeName
+}
+
+func (c *REST) GetConnectionInfo(ctx api.Context, nodeName string) (*client.ConnectionInfo, error) {
+	hostname, port, err := c.getKubeletHostPortPair(nodeName)
 	if err != nil {
-		return "", 0, nil, err
+		return nil, err
 	}
-	if daemonPort > 0 {
-		return scheme, uint(daemonPort), transport, nil
+	connectionInfo, err := c.connection.GetConnectionInfo(ctx, hostname)
```

Comment:
this might read a little better if we swapped the order:

```
// compute the generic connection info
connectionInfo, err := c.connection.GetConnectionInfo(ctx, nodeName)
if err != nil {
  return nil, err
}

// get any specific host/port info for this node
hostname, port, err := c.getKubeletHostPortPair(nodeName)
if err != nil {
  return nil, err
}
if len(hostname) > 0 {
  connectionInfo.Host = hostname
}
if port > 0 {
  connectionInfo.Port = port
}

return connectionInfo, nil
```